### PR TITLE
Overall winner logic

### PR DESCRIPTION
**Pull request by jennysivapalan**
Determines the overall winner. 
- If the minimum number of wins is achieve the winner is determined otherwise it works out who has the max.
- Deals with draws too. 
- It returns null if no one has reached the minimum number of wins required.

### CODE REVIEW

File: client/src/types/game/game.types.tsx [hunk 1]
```diff
@@ -4,5 +4,5 @@ export type Game = {
   players: Player[];
   totalRounds: number;
   currentRound: number;
-  roundWinners: Player[];
+  roundWinners: string[];
```
Comment (jennysivapalan) [inline client/src/types/game/game.types.tsx:7]: Changed as agreed with @beccarothwell who's done the same in her PR on implementing the round.

**Comment by beccarothwell** [comment]
Notes:

- I've just been going through the logic of this and we may need to take out some of what's in determineGameWinner to put into endRound() in App.tsx (and maybe then rename it slightly), and also export findPlayWithMaxWins to use in endRound() then too.
- The reason is because if there are still rounds to play, and there is no clear winner, then we need to call nextPlayer to start the next round. So we could have something like this in endRound():

```
(if roundWinner.length < totalRounds) { 
//if there are still rounds left to play

  const clearWinner = checkForClearWinner()
  
  if (clearWinner) { 
    //if someone has one more than half the available points
    //end game here and declare winner
  }
  
  else { 
    //play another round
    nextPlayer()
  }
}

else { 
  //we’re on the last round and there is no clear winner
  
  const winner = findPlayWithMaxWins()
  //end game here and declare winner
}
```

Alternatively, we'd have to pass nextPlayer() to determineGameWinner() so that it could run it as a callback? But I don't know if that would work?
 
- Minor point too but I realised that `minRequiredToWinGame` would only be `Math.ceil(totalRounds/ 2)` if there's an odd number of rounds. If there's an even number of rounds it's `Math.ceil(totalRounds +1/ 2)`. 
- We may want to rename `minRequiredToWinGame` too, as it's possible to win the game with fewer points if there are more players. For example, if you had a game with 10 rounds and 4 players, you could win with 4 points, or draw with 3 points. 
- Rather than being the minimum required to win the game `Math.ceil(totalRounds/ 2)` is just one more than half the available points, meaning that there are not enough points remaining for anyone else to win.
- We also need to provide logic for if it's a draw. For instance if it's a game of 10 rounds, with 4 players, and two people score 3 points. Maybe have `findPlayWithMaxWins()` return an array of possible winners? If the array is length 1 we know there's one winner, if it's greater, we know it's a draw?

**Comment by jennysivapalan** [comment]
Good point on number of players, rounds and working on a draw - I will address that, should be doable with the tests.

In terms of endRound - I think we need a variable on the game.type to show if it's in progress or ended (because someone won in 3 rounds for example). That way we won't mix the logic between endRound and determineWinner. I can add that to this PR. What do you think @beccarothwell ?

**Comment by jennysivapalan** [comment]
Omg I've been on a journey implementing the logic for determining winner. But think I've solved it and it does the logic for draws too. @beccarothwell would you be okay to re-review this. I will update the other PR on winner-component to deal with draws tomorrow or Monday. 
Phew.

**Comment by beccarothwell** [comment]
> Good point on number of players, rounds and working on a draw - I will address that, should be doable with the tests.
> 
> In terms of endRound - I think we need a variable on the game.type to show if it's in progress or ended (because someone won in 3 rounds for example). That way we won't mix the logic between endRound and determineWinner. I can add that to this PR. What do you think @beccarothwell ?

Missed this comment, but yes agreed. Will add this as part of starting the game, and another PR can then be added to update the win logic to use the isGameActive flag to reset the game.

**Review by beccarothwell — APPROVED** [review]
Looks good, but conflicts will need to be resolved before it can be merged.

**Comment by jennysivapalan** [comment]
I had trouble solving the merge conflicts here, I cuold solve the ones listed above but then I had divergent branches (not sure why) and solving those conflicts wasn't working.

In the end I created a new branch off main and copied the win-logic over. So closing this. In favour of https://github.com/TeamTopTrumps/top-trumps-game/pull/32

**Pull request closed without merging**